internal/lsp/watcher/global_watcher.go

```diff
@@ -0,0 +1,450 @@
+	"github.com/charmbracelet/crush/internal/csync"
+	"github.com/charmbracelet/crush/internal/lsp/protocol"
+	"github.com/fsnotify/fsnotify"
+	ignore "github.com/sabhiram/go-gitignore"
+)
+
+// globalWatcher manages a single fsnotify.Watcher instance shared across all LSP clients.
+//
+// IMPORTANT: This implementation only watches directories, not individual files.
+// The fsnotify library automatically provides events for all files within watched
+// directories, making this approach much more efficient than watching individual files.
+//
+// Key benefits of directory-only watching:
+// - Significantly fewer file descriptors used
+// - Automatic coverage of new files created in watched directories
+// - Better performance with large codebases
+// - fsnotify handles deduplication internally (no need to track watched dirs)
+type globalWatcher struct {
+	watcher *fsnotify.Watcher
+
+	// Map of workspace watchers by client name
+	watchers *csync.Map[string, *WorkspaceWatcher]
+
+	// Single workspace root directory for ignore checking
+	workspaceRoot string
+
+	// Debouncing for file events (shared across all clients)
+	debounceTime time.Duration
+	debounceMap  *csync.Map[string, *time.Timer]
+
+	// Context for shutdown
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	// Wait group for cleanup
+	wg sync.WaitGroup
+}
+
+// getGlobalWatcher returns the singleton global watcher instance
+var getGlobalWatcher = sync.OnceValue(func() *globalWatcher {
+	ctx, cancel := context.WithCancel(context.Background())
+	gw := &globalWatcher{
+		watchers:     csync.NewMap[string, *WorkspaceWatcher](),
+		debounceTime: 300 * time.Millisecond,
+		debounceMap:  csync.NewMap[string, *time.Timer](),
+		ctx:          ctx,
+		cancel:       cancel,
+	}
+
+	// Initialize the fsnotify watcher
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		slog.Error("Failed to create global file watcher", "error", err)
+		return gw
+	}
+
+	gw.watcher = watcher
+
+	// Start the event processing goroutine
+	gw.wg.Add(1)
+	go gw.processEvents()
+
+	return gw
+})
+
+// register registers a workspace watcher with the global watcher
+func (gw *globalWatcher) register(name string, watcher *WorkspaceWatcher) {
+	gw.watchers.Set(name, watcher)
+	slog.Debug("Registered workspace watcher", "name", name)
+}
+
+// unregister removes a workspace watcher from the global watcher
+func (gw *globalWatcher) unregister(name string) {
+	gw.watchers.Del(name)
+	slog.Debug("Unregistered workspace watcher", "name", name)
+}
+
+// watch adds a workspace to be watched.
+// Note: We only watch directories, not individual files. fsnotify automatically provides
+// events for all files within watched directories. Multiple calls with the same workspace
+// are safe since fsnotify handles directory deduplication internally.
+func (gw *globalWatcher) watch(workspacePath string) error {
+	cfg := config.Get()
+	slog.Debug("Adding workspace to global watcher", "path", workspacePath)
```

Comment:
What about prefixing all `slog.Debug` calls with `lsp watcher:`?

```suggestion
	slog.Debug("lsp watcher: adding workspace to global watcher", "path", workspacePath)
```